File: Cpp/01. Elevator System/ElevatorSystem.py
import heapq

class Direction:
    UP = "UP"
    DOWN = "DOWN"
    IDLE = "IDLE"

class Elevator:
    def __init__(self):
        self.current_floor = 0
        self.direction = Direction.IDLE
        self.is_moving = False

    def move_up(self):
        self.current_floor += 1
        self.direction = Direction.UP
        self.is_moving = True
        print(f"Elevator moving up to floor: {self.current_floor}")

    def move_down(self):
        self.current_floor -= 1
        self.direction = Direction.DOWN
        self.is_moving = True
        print(f"Elevator moving down to floor: {self.current_floor}")

    def stop(self):
        self.direction = Direction.IDLE
        self.is_moving = False
        print(f"Elevator stopped at floor: {self.current_floor}\n")

    def should_move_up(self, floor):
        return self.current_floor < floor

    def should_move_down(self, floor):
        return self.current_floor > floor

class ElevatorController:
    def __init__(self):
        self.elevator = Elevator()
        self.up_requests = []  # Min-heap for UP requests
        self.down_requests = []  # Max-heap for DOWN requests (inverted values for max behavior)
        self.processed_requests = set()
        self.destination_floors = {}

    def add_request(self, floor, direction, destination_floor):
        if floor in self.processed_requests:
            return
        self.processed_requests.add(floor)
        self.destination_floors[floor] = destination_floor

        if direction == Direction.UP:
            heapq.heappush(self.up_requests, floor)  # Min-heap
        elif direction == Direction.DOWN:
            heapq.heappush(self.down_requests, -floor)  # Max-heap (negative values)

    def process_requests(self):
        while self.up_requests or self.down_requests:
            if self.elevator.direction == Direction.IDLE:
                self.elevator.direction = Direction.UP

            if self.elevator.direction == Direction.UP:
                self.handle_up_requests()
                if not self.up_requests and self.down_requests:
                    self.elevator.direction = Direction.DOWN

            elif self.elevator.direction == Direction.DOWN:
                self.handle_down_requests()
                if not self.down_requests and self.up_requests:
                    self.elevator.direction = Direction.UP

    def handle_up_requests(self):
        while self.up_requests and self.elevator.direction == Direction.UP:
            next_floor = heapq.heappop(self.up_requests)
            self.move_to_floor(next_floor)

    def handle_down_requests(self):
        while self.down_requests and self.elevator.direction == Direction.DOWN:
            next_floor = -heapq.heappop(self.down_requests)  # Convert back to positive
            self.move_to_floor(next_floor)

    def move_to_floor(self, floor):
        while self.elevator.current_floor != floor:
            if self.elevator.should_move_up(floor):
                self.elevator.move_up()
            elif self.elevator.should_move_down(floor):
                self.elevator.move_down()

        self.elevator.stop()
        dest_floor = self.destination_floors.get(floor)
        if dest_floor is not None:
            self.add_request(dest_floor, Direction.UP if dest_floor > floor else Direction.DOWN, dest_floor)
        self.processed_requests.discard(floor)

# Main function to simulate elevator requests
def main():
    controller = ElevatorController()

    # Simulate requests
    controller.add_request(3, Direction.UP, 7)
    controller.add_request(1, Direction.DOWN, 0)
    controller.add_request(5, Direction.UP, 6)
    controller.add_request(2, Direction.DOWN, 1)

    # Process all requests
    controller.process_requests()

# Run the main function
if __name__ == "__main__":
    main()

File: Cpp/01. Elevator System/ElevatorSystem.js
class Elevator {
  constructor() {
    this.currentFloor = 0;
    this.direction = "IDLE";
    this.isMoving = false;
  }

  moveUp() {
    this.currentFloor++;
    this.direction = "UP";
    this.isMoving = true;
    console.log(`Elevator moving up to floor: ${this.currentFloor}`);
  }

  moveDown() {
    this.currentFloor--;
    this.direction = "DOWN";
    this.isMoving = true;
    console.log(`Elevator moving down to floor: ${this.currentFloor}`);
  }

  stop() {
    this.direction = "IDLE";
    this.isMoving = false;
    console.log(`Elevator stopped at floor: ${this.currentFloor}\n`);
  }

  shouldMoveUp(floor) {
    return this.currentFloor < floor;
  }

  shouldMoveDown(floor) {
    return this.currentFloor > floor;
  }
}

class ElevatorController {
  constructor() {
    this.elevator = new Elevator();
    this.upRequests = [];
    this.downRequests = [];
    this.processedRequests = new Set();
    this.destinationFloors = new Map();
  }

  addRequest(floor, direction, destinationFloor) {
    if (this.processedRequests.has(floor)) return;
    this.processedRequests.add(floor);
    this.destinationFloors.set(floor, destinationFloor);

    if (direction === "UP") {
      this.upRequests.push(floor);
      this.upRequests.sort((a, b) => a - b); // Min-heap simulation
    } else if (direction === "DOWN") {
      this.downRequests.push(floor);
      this.downRequests.sort((a, b) => b - a); // Max-heap simulation
    }
  }

  processRequests() {
    while (this.upRequests.length > 0 || this.downRequests.length > 0) {
      if (this.elevator.direction === "IDLE") {
        this.elevator.direction = "UP";
      }

      if (this.elevator.direction === "UP") {
        this.handleUpRequests();
        if (this.upRequests.length === 0 && this.downRequests.length > 0) {
          this.elevator.direction = "DOWN";
        }
      } else if (this.elevator.direction === "DOWN") {
        this.handleDownRequests();
        if (this.downRequests.length === 0 && this.upRequests.length > 0) {
          this.elevator.direction = "UP";
        }
      }
    }
  }

  handleUpRequests() {
    while (this.upRequests.length > 0 && this.elevator.direction === "UP") {
      let nextFloor = this.upRequests.shift();
      this.moveToFloor(nextFloor);
    }
  }

  handleDownRequests() {
    while (this.downRequests.length > 0 && this.elevator.direction === "DOWN") {
      let nextFloor = this.downRequests.shift();
      this.moveToFloor(nextFloor);
    }
  }

  moveToFloor(floor) {
    while (this.elevator.currentFloor !== floor) {
      if (this.elevator.shouldMoveUp(floor)) {
        this.elevator.moveUp();
      } else if (this.elevator.shouldMoveDown(floor)) {
        this.elevator.moveDown();
      }
    }
    this.elevator.stop();
    let destFloor = this.destinationFloors.get(floor);
    this.addRequest(destFloor, destFloor > floor ? "UP" : "DOWN", destFloor);
    this.processedRequests.delete(floor);
  }
}

// Main function to simulate requests
function main() {
  const controller = new ElevatorController();

  // Simulate requests
  controller.addRequest(3, "UP", 7);
  controller.addRequest(1, "DOWN", 0);
  controller.addRequest(5, "UP", 6);
  controller.addRequest(2, "DOWN", 1);

  // Process all requests
  controller.processRequests();
}

// Run the main function
main();
